scalableminds/webknossos
# Neuroglancer Precomputed Mesh

### URL of deployed dev instance (used for testing):
- https://neuroglancermesh.webknossos.xyz

### Steps to test:
- Test previously created meshes. They should still work even though the protocol has changed (I only tested l4_sample meshfile_4_4_1)
- Test neuroglancer precomputed meshes, I tested:
	- gs://fafb-ffn1-20190805/segmentation/
	- gs://neuroglancer-janelia-flyem-hemibrain/v1.0/segmentation
- Check that meshes can be loaded, downloaded (frontend, backend via route), and align with data
- Render animation does not work for neuroglancer precomputed meshes currently, since the worker does not supply the correct infos (and it does not get these infos from the frontend), see https://github.com/scalableminds/webknossos/issues/8597

### TODOs:
	- [x] Adapt new protocol
	- [x] Move all neuroglancer precomputed mesh handling to different service?
	- [x] ~~Add enum for mesh types~~
		- Replaced the magic values with constants, I think that way its fine for now.
	- [x] ??? Maybe change API to have a mesh object instead of different keys for name, path, type
	- [x] Support private meshes (using credentials?)
		- not for now (since we do not know if these exist)
	- [x] Remove last TODO in DSFullMeshService
		- mapping not supported
	- [x] Can some sharding stuff be unified between array and mesh?
	- [x] Remove "get" for shardingSpec lookup
	- [x] ~~Support non sharding meshes?~~
		- Let's not do this for now, since I haven't encountered any non-sharded meshes, wait until someone actually requests it.
	- [x] Caching
	- [x] Extract stuff for Law of Demeter violations
	- [x] Allow meshes to be found somewhere else other than that "mesh"
	- [x] Do we need to send meshFilePath everywhere or do we want to create that path in the datastore each time
		- I'll keep it for now, since otherwise we would need to request the layer info json every time. Although with #8384 we plan to also add references to other special files, so then the meshfile could also be referenced in the datasource properties (/ later in DB) and then we do not need to send it around anymore.
	- [x] STL download
	- [x] Test regular meshes
	- [x] Remove unused stuff from storage case classes (https://scm.slack.com/archives/C5AKLAV0B/p1744642460772369?thread_ts=1742830834.799849&cid=C5AKLAV0B)
	- [x] Clean up code duplication in meshfileservice
	- [x] **Meshes are loaded, but very far away. Why?**
	- [x] Rename Stuff in NeuroglancerPrecomputedMeshService (we do not need neuroglancer precomputed in the name anymore)
	- [x] Meshes are not shown in Render Animation: Probably because of wrong scale?
		- nvm. Render animation still works (at least for meshes were it previously worked)

### Issues:
- fixes #6826 

------
(Please delete unneeded items, merge only when none are left open)
